sympy/solvers/solveset.py

```diff
@@ -1754,6 +1763,32 @@ def _is_logarithmic(f, symbol):
     return rv
 
 
+def _solve_as_lambert(lhs, rhs, symbol, domain):
+    r"""
+    Computes the equations whose solutions will be given in terms
+    of the `LambertW` function. Returns `None` if it fails doing so.
+
+    Example
+    =======
+
+    >>> from sympy.solvers.solveset import _solve_as_lambert
+    >>> from sympy import symbols, exp, S
+    >>> x = symbols('x')
+    >>> _solve_as_lambert(x*exp(x), 1, x, S.Reals)
+    FiniteSet(LambertW(1))
+
+    """
+    poly = lhs.as_poly()
+    gens = _filtered_gens(poly, symbol)
+    try:
+        soln = _solve_lambert(lhs - rhs, symbol, gens)
+        if domain.is_subset(S.Reals):
+            soln = [s for s in soln if not s.has(I)]
+        return FiniteSet(*soln)
+    except NotImplementedError:
+        pass
```

Comment:
The try/except should be around the smallest amount of code possible.

Reply:
If `soln` is used outside of that then it may give an `unbounded` error!

Reply:
I don't know what you mean. You can write this as:
```python
    try:
        soln = _solve_lambert(lhs - rhs, symbol, gens)
    except NotImplementedError:
        return None
    else:
        if domain.is_subset(S.Reals):
            soln = [s for s in soln if not s.has(I)]
        return FiniteSet(*soln)
```

Reply:
I'm not sure about checking `if not s.has(I)`. Why is that there?

Reply:
> I'm not sure about checking if not s.has(I). Why is that there?

To remove the complex solutions.

Reply:
Can we be sure that any expression with `I` is complex?

Reply:
> Can we be sure that any expression with I is complex?

IMO, a complex will definitely have `I` in most of the cases !

Reply:
An expression can be complex without having I and it can also have I without being complex:
```python
In [11]: e = (1 + I)*(1 - I)

In [12]: e.has(I)
Out[12]: True

In [13]: expand(e)
Out[13]: 2
```
I think a better solution is needed than this.

Reply:
Then what about `expand(s).has(I)` in : ( since `s.is_complex` or `s.is_imaginary` don't give effective results)
```
if domain.is_subset(S.Reals):
            soln = [s for s in soln if not s.has(I)]
```

Reply:
No structural check on the expression can work in all cases. Some cases are undecidable e.g. `Symbol('x')` where the symbol could be anything.

There needs to be a way to handle this and that might mean returning an `Intersection(..., Reals)`.

Reply:
> There needs to be a way to handle this and that might mean returning an Intersection(..., Reals).
Yes, that is already present!

Reply:
What about `if domain.is_subset(S.Reals) and len((lhs-rhs).atoms(Symbol)) == 1:` so that it maybe very definite for non symbol expressions!

```
if domain.is_subset(S.Reals) and len((lhs-rhs).atoms(Symbol)) == 1:
            soln = [s for s in soln if not s.has(I)]
        return FiniteSet(*soln)
```

Reply:
How is that any better than just checking `is_real` or returning an intersection?

Reply:
(Edited above) : `s.is_real` will be fine here!
Should I proceed by taking that into consideration?

Reply:
![Screen Shot 2021-03-16 at 9 42 15 PM](https://user-images.githubusercontent.com/57854867/111342574-97869100-86a0-11eb-83e3-c50eb9709439.png)
![Screen Shot 2021-03-16 at 9 42 27 PM](https://user-images.githubusercontent.com/57854867/111342604-9f463580-86a0-11eb-9bf5-36027995db29.png)

Should I proceed with these changes?

Reply:
Yes, I think that looks better